core/src/main/scala/caliban/Value.scala

```diff
@@ -272,3 +287,144 @@ private object ValuePlayJson {
+    }
+
+    val responseEncoder: JsonEncoder[List[(String, ResponseValue)]] = {
+      val obje = JsonEncoder.keyValueChunk[String, ResponseValue]
+      (a: List[(String, ResponseValue)], indent: Option[Int], out: Write) =>
+        obje.unsafeEncode(Chunk.fromIterable(a), indent, out)
+    }
+  }
+
+  object Arr {
+    val decoder: JsonDecoder[InputValue.ListValue] = {
+      val arrd = JsonDecoder.list[InputValue]
+      (trace: List[JsonDecoder.JsonError], in: RetractReader) => InputValue.ListValue(arrd.unsafeDecode(trace, in))
+    }
+
+    val responseDecoder: JsonDecoder[ResponseValue.ListValue] = {
+      val arrd = JsonDecoder.list[ResponseValue]
+      (trace: List[JsonDecoder.JsonError], in: RetractReader) => ResponseValue.ListValue(arrd.unsafeDecode(trace, in))
+    }
+
+    val responseEncoder: JsonEncoder[List[ResponseValue]] =
+      JsonEncoder.list[ResponseValue]
+
+  }
+
+  object Str {
+    val decoder: JsonDecoder[StringValue] = JsonDecoder.string.map(StringValue)
+  }
+
+  object Num {
+    val decoder: JsonDecoder[Value] = (trace: List[JsonDecoder.JsonError], in: RetractReader) => {
+      val bd = BigDecimal(JsonDecoder.bigDecimal.unsafeDecode(trace, in))
+      if (bd.isValidInt) IntValue(bd.toIntExact)
+      else if (bd.isValidLong) IntValue(bd.toLongExact)
+      else bd.toBigIntExact.fold[Value](FloatValue(bd))(IntValue(_))
+    }
+  }
+
+  implicit val inputValueDecoder: JsonDecoder[InputValue] = (trace: List[JsonDecoder.JsonError], in: RetractReader) => {
+    val c = in.nextNonWhitespace()
+    in.retract()
+    (c: @switch) match {
+      case 'n'       => Null.decoder.unsafeDecode(trace, in)
+      case 'f' | 't' => Bool.decoder.unsafeDecode(trace, in)
+      case '{'       => Obj.decoder.unsafeDecode(trace, in)
+      case '['       => Arr.decoder.unsafeDecode(trace, in)
+      case '"'       => Str.decoder.unsafeDecode(trace, in)
+      case '-' | '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' =>
+        Num.decoder.unsafeDecode(trace, in)
+      case c =>
+        throw JsonDecoder.UnsafeJson(JsonDecoder.JsonError.Message(s"unexpected '$c'") :: trace)
+    }
+
+  }
+
+  val inputValueEncoder: JsonEncoder[InputValue] = (a: InputValue, indent: Option[Int], out: Write) =>
+    a match {
+      case value: Value                 => valueEncoder.unsafeEncode(value, indent, out)
+      case InputValue.ListValue(values) => JsonEncoder.list(inputValueEncoder).unsafeEncode(values, indent, out)
+      case InputValue.ObjectValue(fields) =>
+        JsonEncoder.map(JsonFieldEncoder.string, inputValueEncoder).unsafeEncode(fields, indent, out)
+      case InputValue.VariableValue(name) => JsonEncoder.string.unsafeEncode(name, indent, out)
+    }
+
+  val responseValueDecoder: JsonDecoder[ResponseValue] = (trace: List[JsonDecoder.JsonError], in: RetractReader) => {
+    val c = in.nextNonWhitespace()
+    in.retract()
+    (c: @switch) match {
+      case 'n'       => Null.decoder.unsafeDecode(trace, in)
```

Comment:
We need something that low level?

Reply:
I don't know if it is a bug or not, but I couldn't get the `orElse` operator to work when trying to do a regular composition. I found this code in the `ast` subtree in `zio-json` and ported it. On the plus side I think it should be really fast

Reply:
Discussed in the `zio-json` channel on discord. There are two approaches, one is to use the adt that is provided by zio-json which we would then transform into the caliban ADT. According to the experts we already lose many of the security benefits because of the dynamic nature of the queries. If we go with the ADT then we can remove some of this low level code but we sacrifice the remaining performance benefits (gained by directly converting into our ADT). I can go either way on this, I think the performance gain is a definite plus, but the real reason for this PR is convenience for people wanting to use `zio` all the way through their stack.